src/puppetlabs/puppetdb/jdbc.clj

```diff
@@ -134,16 +134,51 @@
    #(and (map? %)
          (valid-jdbc-query? (:results-query %)))))
 
-(defn coerce-to-array
-  ([x] (coerce-to-array vec x))
-  ([convert x]
-   (cond
-     (kitchensink/array? x) (convert x)
-     (isa? (class x) java.sql.Array) (convert (.getArray ^java.sql.Array x))
-     :else x)))
+(defn sql-array? [x]
+  (isa? (class x) java.sql.Array))
+
+(defn convert-any-sql-array [x convert]
+  (if-not (sql-array? x)
+    x
+    (convert (.getArray ^java.sql.Array x))))
+
+(defn any-sql-array->vec [x]
+  (convert-any-sql-array x vec))
 
 (defn call-with-query-rows
   "Calls (f rows), where rows is a lazy sequence of rows generated
+  from within a transaction.  The sequence is backed by an active
+  database cursor which will be closed when f returns.  Cancels the
+  query if f throws an exception.  The option names that correspond to
+  jdbc/result-set-seq options will affect the rows produced as they do
+  for that function.  So for example, when as-arrays? is logically
+  true, the result rows will be vectors, not maps, and the first
+  result row will be a vector of column names."
+  ([query f] (call-with-query-rows query {} f))
+  ([[sql & params]
+    {:keys [as-arrays? identifiers qualifier read-columns] :as opts}
+    f]
+   (with-db-transaction []
+     (with-open [stmt (.prepareStatement ^Connection (:connection *db*) sql)]
+       (doall (map-indexed (fn [i param] (.setObject stmt (inc i) param))
+                           params))
+       (.setFetchSize stmt 500)
+       (with-open [rset (.executeQuery stmt)]
+         (try
+           (f (sql/result-set-seq rset opts))
+           (catch Exception e
+             ;; Cancel the current query
+             (.cancel stmt)
+             (throw e))))))))
+
+;; Since we only have a couple of array columns, we may want to
```

Comment:
I'm ok deprecating this if you think it's best, but I'd prefer we deprecated rather than leave a note asking that question.